Utils

kashifch opened the pull request:
@benpatterson 

I am breaking the PR in smaller chunks.
Here is the first installments
It basically consists of Utils
I have made following changes
Added unit tests to most of the methods
Brought all regex functions from across tests  in Utils and added unit tests for these
The date conversion function is now generic
I have removed the Email APIs I was using previously, now I am creating new users and reading emails using the python imap plus gmail aliases
The course ids and numbers are now extracted using opaque keys

Things which are not changed
I haven't yet changed the skip_clean_up, as I have commented earlier, I am looking into using the api for un-enrolling multiple users from course, if it works I will remove the function, otherwise we could look into an alternate.

I am still clicking the checkboxes/option buttons using javascript as I experience flakiness when I use the selenium click, if that is unacceptable I can look into other options

Muddasser commented:
Can one of the admins verify this patch?

benpatterson commented on regression/pages/common/tests/test_consts.py:
```diff
@@ -0,0 +1,311 @@
+        Enrollment code for professional seat in Internet of Things: Roadmap
+        to a Connected World
+        Code,Redemption URL
+        AWMHEUSMXYZDJXRN,https://payments.mitprofessionalx-stage.mit.edu/
```
Are these real keys and values? Just a reminder that this is open source...

kashifch replied:
Yes I guess I missed renaming this, will do for all instances

benpatterson commented on regression/pages/common/tests/test_utils.py:
```diff
@@ -0,0 +1,254 @@
+            required_format:
+            result:
+
+        Returns:
```
You don't need the `Returns` info in the docstrings for unit tests

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+    """
+
+    def __init__(self):
+        self.mail = imaplib.IMAP4_SSL('imap.gmail.com')
```
The client should be a variable you set somewhere, whether it's the top of the file or in the constants file. I'll re-read your arguments on why we're not using a test service.... I don't like relying on a third party for such a critical part of tests.

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+    """
+    Get and return yesterday's date in dd-mmm-yyyy format
+    """
+    return (datetime.date.today() - datetime.timedelta(1)).strftime("%d-%b-%Y")
```
is timezone important? I'm assuming yes and wondering how that plays into this.

kashifch replied:
the emails are generated during the tests so there is a lot of buffer for time zone differences

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+        return 'Coupons not found'
+
+
+def skip_cleanup_if_test_passed():
```
Cleanup should be idempotent. You should be able to run it whether the test passed or failed. Its job is to ensure a given end-state.

kashifch replied:
I have created a cleanup which uses the enrollment apis to un-enroll user(s) from course so that should take care of it. I am removing this

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+        page:
+        selectors_and_values_dict:
+    """
+    for key, value in selectors_and_values_dict.iteritems():
```
do you need `iteritems()`?

kashifch replied:
I am looping over the dictionary, if there is a better alternate please let me know.

benpatterson replied:
Sorry what I meant was, can you accomplish the same thing if you remove `.iteritems()` in this statement.

kashifch replied:
it gives "too many value to unpack error" without iteritems()

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+    # pylint: disable=unnecessary-lambda
+    page.wait_for_element_visibility(css_selector, 'wait for target checkbox')
+    page_elem = page.q(css=css_selector).results[0]
+    if not page_elem.is_selected():
```
So this will just move on if a checkbox is already checked? I would worry that this would hide a bug.

kashifch replied:
agreed I will add a check

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+        )
+
+
+def remove_spaces_from_list_elements(list_with_spaces):
```
I don't understand why you would need this?

kashifch replied:
There is some odd behavior between browsers, when reading the countries lists it seems chrome add some extra spaces in a couple of countries which fails the tests in chrome while passing these in firefox, so I am removing all spaces before comparing the lists

benpatterson replied:
I think you should more-fully understand why Chrome is doing this. I think if we create a workaround for poor behavior, it could hide a bug. Do you have any documentation or links describing the problem? I suppose I'm not convinced that we have a bug in our own app here.

kashifch replied:
@benpatterson 

I am not sure it is a bug or not as it is not impacting any functionality, however if you open the profile page in firefox and chrome and use inspect element on countries list you will see that in firefox Falkland is shown as Falkland Islands [Malvinas] but in chrome it has 2 spaces between island and [Malvinas], it is like Falkland Islands  [Malvinas].

Do you think I should log a bug about it?

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+    return [x.replace(' ', '') for x in list_with_spaces]
+
+
+def get_text_against_page_elements(page, css_selectors):
```
nit: s/against/from

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+    list_of_text = []
+    for css_selector in css_selectors:
+        list_of_text.append(page.q(css=css_selector).text[0])
+    return list_of_text
```
I would expect you to return a dict instead of a list, because I'm guessing that the association with the text and where you got the text from is important.

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+    return list_of_text
+
+
+def extract_mmm_dd_yyyy_date_string_from_text(original_string):
```
Is there a reason you can't just cast a string as a `datetime` object?

kashifch replied:
I am just extracting the date from a text chunk, I guess I should rename params and name to make it clear

benpatterson commented on regression/pages/common/tests/test_consts.py:
```diff
@@ -0,0 +1,311 @@
+
+EXTRACT_NUMERICAL_VALUE = {
+    'normal_string_with_dollar_sign': {
+        "price_strings": '$999',
```
You want other strings here. I see one further below, but there's also `$9999.99`, `$.99`, `$0`, `$99,999`.

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+        return 'No numerical value found in search string'
+
+
+def substring_from(original_string, delim):
```
I'm wondering why you wouldn't just say something like `"hi" in "hello hi there"`. I'm not sure why you would want to build a helper function here.

benpatterson commented on regression/pages/common/utils.py:
```diff
@@ -0,0 +1,496 @@
+        return 'Target substring not found'
+
+
+def get_course_key_from_asset(asset_string):
```
Is this method and the one below it...are they not available in the opaque keys library?

kashifch replied:
removing

benpatterson commented on regression/pages/whitelabel/const.py:
```diff
@@ -0,0 +1,327 @@
+CYBERSOURCE_CHECKOUT_URL = \
+    u'https://testsecureacceptance.cybersource.com/checkout'
+
+PROF_COURSE_ID = {
```
Is it ok for us to expose professional course names, IDs in public?

kashifch replied:
I will change courses with test courses

benpatterson replied:
If we need to test actual courses, that is ok. I just think we would remove them from the code and pass it in as a variable (for example).

benpatterson commented on regression/pages/whitelabel/const.py:
```diff
@@ -0,0 +1,327 @@
+VISUAL_USER_EMAIL = 'wl_visual_test01@yopmail.com'
+
+# Countries list
+COUNTRIES_WITH_SPACES = [
```
Is there a reason you can't pull this out of a library somewhere?

kashifch replied:
ok, found a libray

benpatterson commented on regression/pages/whitelabel/const.py:
```diff
@@ -0,0 +1,327 @@
+DIFF_THRESHOLD = 5000
+
+# Timeouts
+DEFAULT_TIMEOUT = 30
```
These should be further up

benpatterson commented on regression/pages/whitelabel/const.py:
```diff
@@ -0,0 +1,327 @@
+LANGUAGES = [x.replace(' ', '') for x in LANGUAGES_WITH_SPACES]
+
+# Default Threshold for visual difference
+DIFF_THRESHOLD = 5000
```
I see the point but think this is not a smart approach. The diff threshold is based on number of pixels. So, you could have a very small image that is a 0% match, but if it's less than 5000 in size, it will always pass.

kashifch commented:
@benpatterson 
Thanks for all the help, I wasn't able to give it proper time in past few days but I have started the work and I think I will finish all changes and submit these by Monday.

I just need your input on storing sensitive data in environment vars, the tests are currently running on 3 sites, pretty soon there will be 6, we would have multiple courses for all these sites especially to test dynamic coupons. 

using 20-30 environment vars doesn't look like a very good option, is there any other way we could hide the data here? It's just a wild idea but can we put all this data in a password protected private repo and then use it in our tests? or any other way of doing it?

kashifch commented:
@benpatterson 

Here are most of the modifications.
Please see if the unittests look better.

The countries and languages lists are unchanged as the third party library (pycountry) has too many differences and there is another library but that is django based and we would need to create a django project to use it.

I have removed the EMAIL class and API class from utils as I have added some more apis. I will share it in next PR.

benpatterson commented on regression/pages/whitelabel/const.py:
```diff
@@ -0,0 +1,335 @@
+    u'Czech Republic', u'Denmark', u'Djibouti', u'Dominica',
+    u'Dominican Republic', u'Ecuador', u'Egypt',
+    u'El Salvador', u'Equatorial Guinea', u'Eritrea', u'Estonia',
+    u'Ethiopia', u'Falkland Islands [Malvinas]',
```
I can't find our comments on it, but you had mentioned that Chrome displays this with two spaces before `[Malvinas]` yet firefox with one space. It is cosmetic, but it sounds like a bug. I wouldn't file one yet, though, I would ask a developer for his/her opinion.

benpatterson replied:
I think installing a library is much better than a list that you're going to have to maintain and fix over time. (Unless you're attempting to test the library itself, but since it's a 3rd-party library, those kinds of tests should exist on that library.) It is more DRY to re-use that library rather than constructing our own list. It's true that we'll be installing more into the project, but for the most part an installed library is not a big deal.

Taking a step back though, why do you need all the countries?

kashifch replied:
we had some instances in the past when a couple of countries got missed, I think we are doing some customization on django library before displaying it and sometimes during the updates or anything a couple of countries got missed from the list and the issue was notified by white label customers, so we have added a check, in this case we are always alerted whenever there is a change in the list.

Ok I will use the django library and see what steps will be needed in automating the creation of django project plus we would need to do the cutomization that is done in platform, I will just cpy paste the functions from there, but can we do it in a later PR, I have lots of functional changes to focus on and I would rather have some tasks in a to do list.

benpatterson commented:
@kashifch Could you explain why we cannot use the same source for countries as we do in production? What are we testing with it?

benpatterson commented:
I also think the unit tests are more readable now; thank you for moving the data to live with the testcases.

benpatterson commented:
I don't understand. What do you mean "creation of the django project?"

On platform, what customization do we do of this list?

sent via Android

On Oct 26, 2016 6:54 AM, "Kashif Chaudhry" <notifications@github.com> wrote:

> *@kashifch* commented on this pull request.
> ------------------------------
>
> In regression/pages/whitelabel/const.py
> <https://github.com/edx/edx-e2e-tests/pull/88>:
>
> > +    u'Belize', u'Benin', u'Bermuda', u'Bhutan', u'Bolivia',
> +    u'Bonaire, Sint Eustatius and Saba',
> +    u'Bosnia and Herzegovina', u'Botswana', u'Bouvet Island', u'Brazil',
> +    u'British Indian Ocean Territory',
> +    u'Brunei', u'Bulgaria', u'Burkina Faso', u'Burundi', u'Cabo Verde',
> +    u'Cambodia', u'Cameroon', u'Canada',
> +    u'Cayman Islands', u'Central African Republic', u'Chad', u'Chile',
> +    u'China', u'Christmas Island',
> +    u'Cocos (Keeling) Islands', u'Colombia', u'Comoros', u'Congo',
> +    u'Congo (the Democratic Republic of the)',
> +    u'Cook Islands', u'Costa Rica', u"C\xf4te d'Ivoire", u'Croatia', u'Cuba',
> +    u'Cura\xe7ao', u'Cyprus',
> +    u'Czech Republic', u'Denmark', u'Djibouti', u'Dominica',
> +    u'Dominican Republic', u'Ecuador', u'Egypt',
> +    u'El Salvador', u'Equatorial Guinea', u'Eritrea', u'Estonia',
> +    u'Ethiopia', u'Falkland Islands [Malvinas]',
>
> we had some instances in the past when a couple of countries got missed, I
> think we are doing some customization on django library before displaying
> it and sometimes during the updates or anything a couple of countries got
> missed from the list and the issue was notified by white label customers,
> so we have added a check, in this case we are always alerted whenever there
> is a change in the list.
>
> Ok I will use the django library and see what steps will be needed in
> automating the creation of django project plus we would need to do the
> cutomization that is done in platform, I will just cpy paste the functions
> from there, but can we do it in a later PR, I have lots of functional
> changes to focus on and I would rather have some tasks in a to do list.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-e2e-tests/pull/88>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6WxunbDUvcfGlsVXlg5HznOZhMEkOHks5q3zF8gaJpZM4KWd7r>
> .
>

kashifch commented:
there you go, removed the lists, we will get back to the library thing later but for now I guess we can dispense with the lists

kashifch commented:
@benpatterson , done a little more digging and it seems we can use the django countries without creating a project, we just need some setting, here is the piece of code that is creating an exact replica of prod list

from django_countries import countries
from django.conf import settings
from django.core.wsgi import get_wsgi_application

settings.configure()
application = get_wsgi_application()

code_and_name = list(countries)
code_and_name = map(list, zip(*code_and_name))
COUNTRIES = code_and_name[1]
find_index = COUNTRIES.index('Kiribati')
COUNTRIES.insert(find_index + 1, 'Kosovo')

currently the only customization is addition of "Kosovo" in the list.

Also looking at the django countries list I found out that the django countries list has two spaces in Falkland Islands  [Malvinas], so it seems the chrome is displaying the name correctly, firefox is removing one space. 

I will now take a look at the languages

kashifch commented:
hmm, it seems that languages will pose an issue, in platform a custom list of languages is defined based on http://loc.gov/standards/iso639-2/ISO-639-2_utf-8.txt, the list is defined at lms/env/common.py. It seems that we are using around 186 languages from the original list of around 485.

benpatterson commented:
@kashifch thank you for digging.

Again, what are you trying to test wherein you need all these countries? Specifically, could you describe the test cases or the functionality that is being tested?

kashifch commented:
Just checking that all the countries and languages are present in the dropdown of profile page, I don't remember the exact ticket (it was about some countries not loading in the drop down) cause the issue happened a long time ago and Chris Dodge thought it would be good idea to check nothing is missing, if this scenario seems unnecessary to you we can just remove the test case. I haven't seen it failing in the last year except when the spellings for a couple of countries changed and that was legitimate change

kashifch commented:
do you think this PR is ready to be merged, I will create the next one containing API classes and email class once it is merged

benpatterson commented:
Ok thanks for the context. 

Checking that nothing is missing....it is overkill to check every single line. You can accomplish the same end by doing a count of the total number of items in the dropdown and making sure it's the right number. 

If you want to be sure the country names are rendered correctly, you could select some of the names (like tricky ones with odd characters) and ensure those are in the dropdown as a separate test case.

I'd go that route. Again, keeping the list of countries up to date will be a maintenance headache and not worth the cost, so the idea is you think of another way of tackling the problem.

kashifch commented:
apart from the planned scenarios we do add test cases for every scenario that has ever failed on white label sites, like recently the ora file upload was not working on white label so we will be adding a test case for it, also there was an issue that files in static folders were missing the hash after name so we have added a check for it too

kashifch commented:
hmm ok, I guess we can store the number of countries and languages in a variable plus have a smaller list of some sample items to test for name formats, that way we don't have to keep a big list or install libraries

benpatterson commented:
it was a bit unclear to me what you were responding to, but, in terms of creating new test cases when a failure is observed, yes that makes sense. again the implementation details of those test cases are where there is room for unmaintainable test code.

kashifch commented:
sorry for the disconnect but you got the gist of it and I totally agree with you.

On a different note can you take a look at https://build.testeng.edx.org/view/All/job/microsites-staging-tests/ whenever you have some time. I configured it to run using this repo and it worked successfully once but now it is throwing browser errors

benpatterson commented:
we should take that discussion offline. i'll send an e-mail.

kashifch commented:
@benpatterson 

I think all the requested changes have been made, if you could give a thumbs up I will merge this and share another PR for review

benpatterson commented:
👍 from me

benpatterson commented:
please rebase before merging.

kashifch commented:
Thanks
